File: node/packages/fluent-cards/src/builders/inputs/InputToggleBuilder.ts
import { InputLabelPosition, InputStyle, Spacing } from '../../enums.js';
import type { InputToggle, AdaptiveElement } from '../../models.js';

/** Fluent builder for {@link InputToggle} elements. */
export class InputToggleBuilder {
  private readonly input: InputToggle = { type: 'Input.Toggle', id: '', title: '' };

  /** Sets the unique identifier (required for form submission). @param id The unique identifier. @returns The builder instance for method chaining. */
  withId(id: string): this { this.input.id = id; return this; }
  /** Sets the required toggle label displayed beside the checkbox. @param title The toggle label. @returns The builder instance for method chaining. */
  withTitle(title: string): this { this.input.title = title; return this; }
  /** Sets the additional descriptive label displayed above the toggle. @param label The label text. @returns The builder instance for method chaining. */
  withLabel(label: string): this { this.input.label = label; return this; }
  /** Sets the initial value of the toggle. @param value The initial value. @returns The builder instance for method chaining. */
  withValue(value: string): this { this.input.value = value; return this; }
  /** Sets the value submitted when the toggle is on. @param valueOn The on value. @returns The builder instance for method chaining. */
  withValueOn(valueOn: string): this { this.input.valueOn = valueOn; return this; }
  /** Sets the value submitted when the toggle is off. @param valueOff The off value. @returns The builder instance for method chaining. */
  withValueOff(valueOff: string): this { this.input.valueOff = valueOff; return this; }
  /** Sets whether the toggle label wraps. @param wrap True to allow wrapping. @returns The builder instance for method chaining. */
  withWrap(wrap = true): this { this.input.wrap = wrap; return this; }
  /** Sets whether a value is required. @param isRequired True if required. @returns The builder instance for method chaining. */
  withIsRequired(isRequired = true): this { this.input.isRequired = isRequired; return this; }
  /** Sets the error message shown when validation fails. @param errorMessage The error message. @returns The builder instance for method chaining. */
  withErrorMessage(errorMessage: string): this { this.input.errorMessage = errorMessage; return this; }
  /** Sets the spacing above the element. @param spacing The spacing value. @returns The builder instance for method chaining. */
  withSpacing(spacing: Spacing): this { this.input.spacing = spacing; return this; }
  /** Sets whether a separator line is displayed above the element. @param separator True to show a separator. @returns The builder instance for method chaining. */
  withSeparator(separator = true): this { this.input.separator = separator; return this; }
  /** Sets whether the element is visible. @param isVisible True to show the element. @returns The builder instance for method chaining. */
  withIsVisible(isVisible: boolean): this { this.input.isVisible = isVisible; return this; }
  /** Sets the height of the element. @param height The height ('auto' or 'stretch'). @returns The builder instance for method chaining. */
  withHeight(height: string): this { this.input.height = height; return this; }
  /** Sets the fallback behavior when the element is unsupported. @param fallback The fallback value ('drop' or an element). @returns The builder instance for method chaining. */
  withFallback(fallback: 'drop' | AdaptiveElement): this { this.input.fallback = fallback; return this; }
  /** Sets the feature requirements for the element. @param key The feature name. @param version The minimum required version. @returns The builder instance for method chaining. */
  withRequires(key: string, version: string): this { this.input.requires = { ...this.input.requires, [key]: version }; return this; }
  /** Sets whether content should be laid out right-to-left. @param rtl True for RTL. @returns The builder instance for method chaining. */
  withRtl(rtl = true): this { this.input.rtl = rtl; return this; }
  /** Sets the label position relative to the input. @param position The label position. @returns The builder instance for method chaining. */
  withLabelPosition(position: InputLabelPosition): this { this.input.labelPosition = position; return this; }
  /** Sets the width of the label. @param width The label width (e.g. '100px' or '30%'). @returns The builder instance for method chaining. */
  withLabelWidth(width: string): this { this.input.labelWidth = width; return this; }
  /** Sets the visual style of the input. @param style The input style. @returns The builder instance for method chaining. */
  withInputStyle(style: InputStyle): this { this.input.inputStyle = style; return this; }

  /** Builds and returns the configured InputToggle element. @returns The configured InputToggle instance. */
  build(): InputToggle { return this.input; }
}

File: python/src/fluent_cards/builders/inputs/input_toggle_builder.py
from __future__ import annotations
from ...enums import Spacing, InputLabelPosition, InputStyle


class InputToggleBuilder:
    """Fluent builder for creating Input.Toggle elements."""

    def __init__(self):
        self._input: dict = {'type': 'Input.Toggle', 'id': '', 'title': ''}

    def with_id(self, id: str) -> InputToggleBuilder:
        """Sets the unique identifier for the Input.Toggle element.

        Args:
            id: The unique identifier used when submitting form data.

        Returns:
            The builder instance for method chaining.
        """
        self._input['id'] = id
        return self

    def with_title(self, title: str) -> InputToggleBuilder:
        """Sets the title text displayed next to the toggle.

        Args:
            title: The toggle label text.

        Returns:
            The builder instance for method chaining.
        """
        self._input['title'] = title
        return self

    def with_label(self, label: str) -> InputToggleBuilder:
        """Sets the label displayed above the toggle.

        Args:
            label: The label text.

        Returns:
            The builder instance for method chaining.
        """
        self._input['label'] = label
        return self

    def with_value(self, value: str) -> InputToggleBuilder:
        """Sets the initial value of the toggle.

        Args:
            value: The default value ('true' or 'false', or custom valueOn/valueOff strings).

        Returns:
            The builder instance for method chaining.
        """
        self._input['value'] = value
        return self

    def with_value_on(self, value_on: str) -> InputToggleBuilder:
        """Sets the value submitted when the toggle is on.

        Args:
            value_on: The value representing the on state.

        Returns:
            The builder instance for method chaining.
        """
        self._input['valueOn'] = value_on
        return self

    def with_value_off(self, value_off: str) -> InputToggleBuilder:
        """Sets the value submitted when the toggle is off.

        Args:
            value_off: The value representing the off state.

        Returns:
            The builder instance for method chaining.
        """
        self._input['valueOff'] = value_off
        return self

    def with_wrap(self, wrap: bool = True) -> InputToggleBuilder:
        """Sets whether the toggle title text wraps.

        Args:
            wrap: True to allow text wrapping.

        Returns:
            The builder instance for method chaining.
        """
        self._input['wrap'] = wrap
        return self

    def with_is_required(self, is_required: bool = True) -> InputToggleBuilder:
        """Sets whether the input field is required.

        Args:
            is_required: True to mark the field as required.

        Returns:
            The builder instance for method chaining.
        """
        self._input['isRequired'] = is_required
        return self

    def with_error_message(self, error_message: str) -> InputToggleBuilder:
        """Sets the error message shown when validation fails.

        Args:
            error_message: The validation error message text.

        Returns:
            The builder instance for method chaining.
        """
        self._input['errorMessage'] = error_message
        return self

    def with_spacing(self, spacing: Spacing) -> InputToggleBuilder:
        """Sets the spacing before the toggle element.

        Args:
            spacing: The spacing value.

        Returns:
            The builder instance for method chaining.
        """
        self._input['spacing'] = spacing.value
        return self

    def with_is_visible(self, is_visible: bool) -> InputToggleBuilder:
        """Sets whether the element is visible."""
        self._input['isVisible'] = is_visible
        return self

    def with_separator(self, separator: bool = True) -> InputToggleBuilder:
        """Sets whether a separator line is drawn above the element."""
        self._input['separator'] = separator
        return self

    def with_height(self, height: str) -> InputToggleBuilder:
        """Sets the height of the element."""
        self._input['height'] = height
        return self

    def with_fallback(self, fallback) -> InputToggleBuilder:
        """Sets the fallback content if the element is unsupported."""
        self._input['fallback'] = fallback
        return self

    def with_requires(self, key: str, version: str) -> InputToggleBuilder:
        """Sets a feature requirement for the element."""
        if 'requires' not in self._input:
            self._input['requires'] = {}
        self._input['requires'][key] = version
        return self

    def with_rtl(self, rtl: bool = True) -> InputToggleBuilder:
        """Sets right-to-left text direction."""
        self._input['rtl'] = rtl
        return self

    def with_label_position(self, position: InputLabelPosition) -> InputToggleBuilder:
        """Sets the position of the label relative to the input."""
        self._input['labelPosition'] = position.value
        return self

    def with_label_width(self, width: str) -> InputToggleBuilder:
        """Sets the width of the label."""
        self._input['labelWidth'] = width
        return self

    def with_input_style(self, style: InputStyle) -> InputToggleBuilder:
        """Sets the visual style of the input."""
        self._input['inputStyle'] = style.value
        return self

    def build(self) -> dict:
        """Builds and returns the configured Input.Toggle dictionary.

        Returns:
            The configured Input.Toggle dictionary.
        """
        return self._input
